File: packages/jexlang-java/src/main/java/com/jexlang/java/functions/MapFuncRegistry.java
package com.jexlang.java.functions;

import com.jexlang.java.types.JexValue;

import java.util.Map;

public final class MapFuncRegistry implements FuncRegistry {
    private final Map<String, FuncImpl> map;

    public MapFuncRegistry(Map<String, FuncImpl> funcMap) {
        this.map = funcMap;
    }

    public void set(String name, FuncImpl fn) { map.put(name, fn); }
    public boolean has(String name) { return map.containsKey(name); }
    public JexValue call(String name, JexValue... args) {
        FuncImpl f = map.get(name);
        if (f == null) throw new RuntimeException("Unknown function: " + name);
        return f.apply(args);
    }
}

File: packages/jexlang-swift/jexlang-swift/eval/functions/MapFuncRegistry.swift
//
//  MapFuncRegistry.swift
//  jexlang-swift
//
//  Created by Vinay Kumar on 04/11/25.
//

import Foundation

public final class MapFuncRegistry: FuncRegistry {
    private var map: [String: FuncImpl]

    public init(_ funcMap: [String: FuncImpl] = [:]) {
        self.map = funcMap
    }

    public func set(_ name: String, _ fn: @escaping FuncImpl) {
        map[name] = fn
    }

    public func getAll() -> [String: FuncImpl] {
        return map
    }

    public func has(_ name: String) -> Bool {
        return map[name] != nil
    }

    public func remove(_ name: String) {
        map.removeValue(forKey: name)
    }

    public func call(_ name: String, _ ctx: EvaluatorContext, _ args: [JexValue]) -> JexValue {
        guard let fn = map[name] else {
            fatalError("Unknown function: \(name)")
        }
        return fn(ctx, args)
    }
}
